src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,217 @@
+use crate::error::{Error, Res};
+use crate::threshold::Ciphertext;
+use hex::encode as hex;
+/// This module implements privacy budget using a redis cache
+use redis::{transaction, Commands, RedisError};
+use rust_elgamal::{DecryptionKey, RistrettoPoint, Scalar};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct BlindedMatchKey(Ciphertext);
+
+#[allow(dead_code)]
+impl BlindedMatchKey {
+    pub fn new(key: DecryptionKey, message: RistrettoPoint, scalar: Scalar) -> Self {
+        let c = key.encryption_key().encrypt_with(message, scalar);
+        Self(c)
+    }
+}
+
+impl ToString for BlindedMatchKey {
+    fn to_string(&self) -> String {
+        let (c0, c1) = self.0.inner();
+        let mut c0_hex = hex(c0.compress().to_bytes());
+        let c1_hex = hex(c1.compress().to_bytes());
+        c0_hex.push_str(&c1_hex);
+        c0_hex
+    }
+}
+
+struct PrivacyBudget {
+    redis_connection: redis::Connection,
+    total_privacy_budget: u32,
+}
+
+trait ToString {
+    fn to_string(&self) -> String;
+}
+
+impl ToString for Ciphertext {
+    fn to_string(&self) -> String {
+        let (c0, c1) = self.inner();
+        let mut c0_hex = hex(c0.compress().to_bytes());
+        let c1_hex = hex(c1.compress().to_bytes());
+        c0_hex.push_str(&c1_hex);
+        c0_hex
+    }
+}
+
+#[allow(dead_code)]
+impl PrivacyBudget {
+    fn new(
+        redis_host_name: &str,
+        redis_password: Option<String>,
+        is_tls: bool,
+    ) -> Res<PrivacyBudget> {
+        //if Redis server needs secure connection
+        let uri_scheme = if is_tls { "rediss" } else { "redis" };
+
+        let redis_conn_url = format!(
+            "{}://:{}@{}",
+            uri_scheme,
+            redis_password.unwrap_or_default(),
+            redis_host_name
+        );
+
+        let redis_connection = redis::Client::open(redis_conn_url)?.get_connection()?;
+
+        Ok(PrivacyBudget {
+            redis_connection,
+            total_privacy_budget: 100,
+        })
+    }
+
+    /// Input : a vector of ciphertexts alongwith budget to be consumed.
+    /// This atomically increases the budget of all of the blinded match keys that are received.
+    /// For those that end up exceeding the budget, the budget is then decremented again,
+    /// but the corresponding match key is discarded and not passed to the output.
+    pub fn get_eligible_packets_with_budget(
```

Comment:
I'm still not particularly happy with this interface.

A consumer of this API might want to feed data in via an iterator (rather than having you call .iter() on a slice).  And they don't necessarily want to have clones of their inputs provided to them.  Returning an iterator that includes a subset of the inputs is closer to what the caller will need.

Reply:
While making the function return an iterator, there are some considerations. Since iterator holds on to the &mut self which is passed on to it, we would need to restrict the lifetime of the packets to that of privacy budget object. Is that an acceptable outcome?